quiche/src/build.rs

```diff
@@ -236,13 +236,13 @@ fn main() {
         let build_dir = format!("{}/build/{}", bssl_dir, build_path);
         println!("cargo:rustc-link-search=native={}", build_dir);
 
-        println!("cargo:rustc-link-lib=static=ssl");
-        println!("cargo:rustc-link-lib=static=crypto");
+        println!("cargo:rustc-link-lib=static:+whole-archive=ssl");
+        println!("cargo:rustc-link-lib=static:+whole-archive=crypto");
```

Comment:
From the Rust release notes:
> (more common) either be reordered to respect dependencies between them (if a depends on b then a should go first and b second)
that is already the case here, as libssl depends on libcrypto, but maybe worth trying to swap the order to see if that fixes the problem as well? If you haven't tried already...

Reply:
It seems that my (admittedly very naive) approach doesn't work.

Fixing the order would be a better solution.

Do we know what exactly the ordering problem is and how to correct it?

Reply:
Reversing the order of the `ssl` and `crypto` lines allows quiche to build (it builds as-is too, but it any attempt to use quiche results in a segfault). However, compiling against it results in an error. Here's the output of attempting to build curl 7.85.0 when quiche has been modified to reverse these 2 lines:
```
configure:29986: checking for quiche options with pkg-config
configure:30000: result: found
configure:30014: -l is -lquiche 
configure:30023: -I is 
configure:30033: -L is 
configure:30043: checking for quiche_connect in -lquiche
configure:30066: x86_64-pc-linux-gnu-gcc -o conftest -O2 -march=native -pipe -fuse-linker-plugin -flto -Wl,-flto -ftree-vectorize -falign-functions=32 -fgraphite-identity -floop-nest-optimize -fno-common -fno-lto -fno-use-linker-plugin -Werror-implicit-function-declaration           -Wl,-O1 -Wl,--as-needed -flto -Wl,--hash-style=gnu -fno-lto -fno-use-linker-plugin          conftest.c -lquiche  -lquiche  -lnghttp2  -lidn2  -lrtmp -lz -lssl -lcrypto  -lssh2 -lssh2  -lssl -lcrypto -lssl -lcrypto  -lzstd -lzstd  -lbrotlidec -lbrotlidec  -lz  >&5
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `HRSS_decap'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `HRSS_encap'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `BIO_get_fd'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_AEAD_id'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `HKDF_expand'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_aead_aes_256_cbc_sha1_tls'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_aead_aes_256_cbc_sha1_tls_implicit_iv'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_KEY_zero'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `HRSS_generate_key'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `BIO_set_fd'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_aead_aes_128_cbc_sha1_tls_implicit_iv'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_hpke_aes_256_gcm'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `HRSS_parse_public_key'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `HRSS_marshal_public_key'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_CTX_zero'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_hpke_aes_128_gcm'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_CTX_cleanup'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_AEAD_aead'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_KEY_public_key'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_CTX_seal'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_KEY_kem'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_hpke_hkdf_sha256'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_hpke_x25519_hkdf_sha256'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `HKDF_extract'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_KDF_id'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `CRYPTO_library_init'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_CTX_setup_sender'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_aead_null_sha1_tls'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_CTX_kdf'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_KEY_cleanup'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_aead_des_ede3_cbc_sha1_tls_implicit_iv'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_CTX_setup_recipient'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_CTX_open'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_CTX_aead'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_aead_aes_128_cbc_sha1_tls'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_hpke_chacha20_poly1305'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_aead_des_ede3_cbc_sha1_tls'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_KEM_id'
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../lib64/libquiche.so: undefined reference to `EVP_HPKE_KEY_copy'
collect2: error: ld returned 1 exit status
```

Reply:
> -lquiche  -lquiche  -lnghttp2  -lidn2  -lrtmp -lz -lssl -lcrypto

One possibility is that `-lquiche` somehow needs to be moved after `-lssl -lcrypto` in the curl build maybe?

Reply:
Also, for curl usually we keep the `boringssl-vendored` / `boringssl-boring-crate` features disabled, as BoringSSL is usually built externally so there's no need for quiche to build it again.